### PR TITLE
Naval units (RepairableNear) units repaired with Repair cursor #12361

### PR DESCRIPTION
**Pull request by rob-v**
Repair cursor now supports also RepairableNear units - ships, submarines - in addition to Repairable vehicles.

### CODE REVIEW

File: OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs [hunk 1]
```diff
@@ -49,17 +49,28 @@ static IEnumerable<Order> OrderInner(World world, MouseInput mi)
 			if (underCursor.Owner != world.LocalPlayer)
 				yield break;
 
+			Actor repairBuilding = null;
+			string orderId = "Repair";
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs]: `var orderId ...`

File: OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs [hunk 2]
```diff
@@ -49,17 +49,28 @@ static IEnumerable<Order> OrderInner(World world, MouseInput mi)
...
+
 			// Test for generic Repairable (used on units).
 			var repairable = underCursor.TraitOrDefault<Repairable>();
-			if (repairable == null)
-				yield break;
+			if (repairable != null)
+				repairBuilding = repairable.FindRepairBuilding(underCursor);
+			else
+			{
+				// Test for generic RepairableNear (used on ships).
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs]: Unnecessary comment.

File: OpenRA.Mods.Common/Traits/RepairableNear.cs [hunk 3]
```diff
@@ -84,5 +85,17 @@ public void ResolveOrder(Actor self, Order order)
...
+		public Actor FindRepairBuilding(Actor self)
+		{
+			var repairBuilding = self.World.ActorsWithTrait<RepairsUnits>()
+				.Where(a => !a.Actor.IsDead && a.Actor.IsInWorld
+					&& a.Actor.Owner.IsAlliedWith(self.Owner) &&
+					info.Buildings.Contains(a.Actor.Info.Name))
+				.OrderBy(p => (self.Location - p.Actor.Location).LengthSquared);
+
+			// Worst case FirstOrDefault() will return a TraitPair<null, null>, which is OK.
+			return repairBuilding.FirstOrDefault().Actor;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/RepairableNear.cs]: If `ActorsWithTrait` returns an empty collection `FirstOrDefault` will return `(TraitPair'2)null`.

Comment (rob-v) [inline OpenRA.Mods.Common/Traits/RepairableNear.cs]: I didn't check this, it is adapted copy from Repairable.FindRepairBuilding(). It contains this comment. Should I check it or can we assume it is correct?

File: OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs (hunk 2 above)
Comment (rob-v) [inline OpenRA.Mods.Common/Orders/RepairOrderGenerator.cs]: There was already comment 'Test for generic Repairable', so I added this one for consistency. Should I remove only mine or both?

File: OpenRA.Mods.Common/Traits/RepairableNear.cs (hunk 3 above)
Comment (rob-v) [inline OpenRA.Mods.Common/Traits/RepairableNear.cs]: repairBuilding is IOrderedEnumerable of TraitPair-s what is Struct, so FirstOrDefault() will never return null, only Actor can be null in such case like the comment says.

**Comment by rob-v** [comment]
Style fixed.

**Review by pchote — APPROVED** [review]

**Comment by atlimit8** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a4fc5ad43c135064d38334dd1d67cefc39a150b3...55da00c151aa4977132a524c85ad0d9575e68796)

**Pull request merged**